scene/resources/mesh_data_tool.cpp

```diff
@@ -446,6 +515,69 @@ void MeshDataTool::set_vertex_meta(int p_idx, const Variant &p_meta) {
 	vertices.write[p_idx].meta = p_meta;
 }
 
+Variant MeshDataTool::get_vertex_custom(int p_idx, int p_cidx) const {
+	ERR_FAIL_INDEX_V(p_idx, vertices.size(), Variant());
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, Variant());
+	uint32_t type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+	if (type > Mesh::ARRAY_CUSTOM_RGBA_HALF) {
+		return vertices[p_idx].custom[p_cidx].fd;
+	} else {
+		Vector<uint8_t> ret;
+		ret.resize(8);
+		memcpy(ret.ptrw(), vertices[p_idx].custom[p_cidx].ud, 8);
+		return ret;
+	}
+}
+
+Error MeshDataTool::set_vertex_custom(int p_idx, int p_cidx, const Color &p_customf, const PackedByteArray &p_customu) {
+	ERR_FAIL_INDEX_V(p_idx, vertices.size(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, ERR_INVALID_PARAMETER);
+
+	if (format & (Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx)) {
+		uint32_t type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+		if (type > Mesh::ARRAY_CUSTOM_RGBA_HALF) {
+			vertices.write[p_idx].custom[p_cidx].fd = p_customf;
+		} else {
+			uint32_t count = MIN(p_customu.size(), 8);
+			memcpy(vertices.write[p_idx].custom[p_cidx].ud, p_customu.ptr(), count);
+		}
+	} else {
+		// default behaviour setting the format to rgba_float
+		vertices.write[p_idx].custom[p_cidx].fd = p_customf;
+		format |= Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx;
+		format |= Mesh::ArrayCustomFormat::ARRAY_CUSTOM_RGBA_FLOAT << (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS);
+	}
+
+	return OK;
+}
+
+int MeshDataTool::has_custom(int p_cidx) const {
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, 0);
+	return format & (Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx);
+}
+
+int MeshDataTool::get_custom_type(int p_cidx) const {
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, -1);
+	ERR_FAIL_COND_V((format & (Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx)) == 0, -1);
+	uint32_t type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+	return type;
+}
+
+Error MeshDataTool::set_custom_type(int p_cidx, uint32_t p_type, bool p_convert) {
+	ERR_FAIL_INDEX_V(p_cidx, Mesh::ARRAY_CUSTOM_COUNT, ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V(p_type >= Mesh::ARRAY_CUSTOM_MAX, ERR_INVALID_PARAMETER);
+	if (p_convert && format & (Mesh::ARRAY_FORMAT_CUSTOM0 << p_cidx)) {
+		uint32_t prev_type = (format >> (Mesh::ARRAY_FORMAT_CUSTOM_BASE + p_cidx * Mesh::ARRAY_FORMAT_CUSTOM_BITS)) & Mesh::ARRAY_FORMAT_CUSTOM_MASK;
+		if (prev_type != p_type) {
+			// TODO: add type conversion for when it's possible
+		}
```

Comment:
perhaps just make an ERR_FAIL() if you change the format when there are vertices.

But I'm getting really scared reading all this code and imagining all the conversion. Hmm... I think we should try something different here.